Support Range in :values

davydog187 opened the pull request:
This PR adds support to prop `:values` such that a range can be used to communicate the possible values.

### Example
```elixir
defmodule MyComponent do
  use Surface.Component
  
  prop age, :integer, values: 1..100, default: 1
  
  ...
end
```

### Considerations
This adds `is_struct/2` as a guard, which was added in Elixir 1.11. I'm not sure which version we're looking to support. If we need to support earlier versions, then I believe I'll need to refactor `validate_opt/7` as the way it currently works only supports rejecting values

### Follow ups
If this PR is accepted, I will also PR to update the docs and the catalogue to support this new feature

miguel-s reviewed (commented):
Looks good to me!

You'll need to update the error message in the exisiting test at `api_test.exs:46`.

davydog187 commented:
@miguel-s done, and changed the minimum supported elixir version to 1.11, and updated the CI matrix to reflect.

Per slack, @msaraiva seems ok with this.

msaraiva commented:
Hi @davydog187!

Thanks for the PR. Do you know if this requires any change in [surface_catalogue](https://github.com/surface-ui/surface_catalogue) and [surface_site](https://github.com/surface-ui/surface_site)? Since both present the list of values in the docs, we need to make sure it doesn't break anything.

![image](https://user-images.githubusercontent.com/457237/116408388-1b3bad80-a809-11eb-9321-e9e202c1d31c.png)

davydog187 commented:
@msaraiva yes it will! I mentioned it in the description, you may have missed it. I'll add those PRs now

msaraiva commented:
@davydog187 yes. I missed it. Sorry. 😁

miguel-s reviewed (approved):
Thanks @davydog187 ❤️

msaraiva commented:
@davydog187 since this one sets the minimum version of Elixir to `v1.11`, I'll leave it out of `v0.4`. We can merge it before `v0.5`, which should not take long as we already have the new changes in the syntax approved and the work has already begun.

davydog187 commented:
I have a similar one coming. No worries

davydog187 commented:
Updated the documentation https://github.com/surface-ui/surface_site/pull/33

msaraiva commented:
Hey @davydog187!

Since we already have an "official" branch for `v0.5`, which is https://github.com/surface-ui/surface/tree/ms-use-custom-tokenizer, it would be nice to merge this there. Can you submit this PR against that branch instead of master? You can do the same for any follow-up PR you might have that should be applied to `v0.5.0`.

Cheers.

davydog187 commented:
You got it!

msaraiva commented:
@davydog187 quick correction. Please use https://github.com/surface-ui/surface/tree/surface-next as the target for any PR that should be applied to `v0.5`.  Sorry for the last-minute change. Thanks.

davydog187 commented:
I probably won't get to it until Monday. No worries, thanks for the update

davydog187 commented:
@msaraiva rebased with `surface-next` and set that branch as the target of this PR.

davydog187 commented:
Once this is merged i'll do the same with my other PR as it will have merge conflicts

msaraiva commented:
@davydog187 thanks! ❤️
